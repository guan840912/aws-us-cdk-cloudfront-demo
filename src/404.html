<h1>This is a custom 404 error page.</h1>