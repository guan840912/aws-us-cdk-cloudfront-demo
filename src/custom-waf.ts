import * as wafv2 from '@aws-cdk/aws-wafv2';
import { Construct } from '@aws-cdk/core';
export class CustomWebAcl extends Construct {
  readonly webAcl:wafv2.CfnWebACL;
  constructor(scope: Construct, id: string) {
    super(scope, id );
    this.webAcl = new wafv2.CfnWebACL(this, 'webACL', {
      name: 'cdk-webACL',
      defaultAction: {
        block: {},
      },
      scope: 'CLOUDFRONT',
      visibilityConfig: {
        sampledRequestsEnabled: true,
        cloudWatchMetricsEnabled: true,
        metricName: 'cdk-webACL',
      },
      rules: [{
        action: {
          allow: {},
        },
        name: 'allow-default-index-and-end-slash',
        priority: 0,
        visibilityConfig: {
          sampledRequestsEnabled: true,
          cloudWatchMetricsEnabled: true,
          metricName: 'CustomWafACL',
        },
        statement: {
          orStatement: {
            statements: [
              {
                byteMatchStatement: {
                  searchString: '.html',
                  fieldToMatch: {
                    uriPath: {},
                  },
                  textTransformations: [
                    {
                      priority: 0,
                      type: 'NONE',
                    },
                  ],
                  positionalConstraint: 'ENDS_WITH',
                },
              },
              {
                byteMatchStatement: {
                  searchString: '/',
                  fieldToMatch: {
                    uriPath: {},
                  },
                  textTransformations: [
                    {
                      priority: 0,
                      type: 'NONE',
                    },
                  ],
                  positionalConstraint: 'ENDS_WITH',
                },
              },
            ],
          },
        },
      }],
    });
  }
};