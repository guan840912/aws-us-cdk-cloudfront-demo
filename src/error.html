<h1>This is an ERROR.</h1>