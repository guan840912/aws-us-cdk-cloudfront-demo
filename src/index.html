<h1>Hello CDK!</h1>