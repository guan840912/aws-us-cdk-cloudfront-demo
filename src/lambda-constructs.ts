import * as path from 'path';
import * as cf from '@aws-cdk/aws-cloudfront';
import * as lambda from '@aws-cdk/aws-lambda';
import * as lambdanj from '@aws-cdk/aws-lambda-nodejs';
import { Construct } from '@aws-cdk/core';


/**
 * Display customized error pages, or mask 4XX error pages, based on where the error originated.
 *
 * see https://aws.amazon.com/tw/blogs/networking-and-content-delivery/customize-403-error-pages-from-amazon-cloudfront-origin-with-lambdaedge/
 */
export class CustomErrorPage extends Construct {
  readonly lambdaFunction: lambda.Version;
  readonly eventType: cf.LambdaEdgeEventType;
  constructor(scope: Construct, id: string) {
    super(scope, id );
    const func = new lambdanj.NodejsFunction(this, 'CustomFunc', {
      runtime: lambda.Runtime.NODEJS_12_X,
      entry: path.join(__dirname, './lambda/cf-custom-error-page/index.ts'),
    });

    this.eventType = cf.LambdaEdgeEventType.ORIGIN_RESPONSE,
    this.lambdaFunction = func.currentVersion;
  }
};

/**
 * Default Directory Indexes in Amazon S3-backed Amazon CloudFront Origins.
 *
 * see use case - see https://aws.amazon.com/tw/blogs/compute/implementing-default-directory-indexes-in-amazon-s3-backed-amazon-cloudfront-origins-using-lambdaedge/
 */
export class DefaultDirIndex extends Construct {
  readonly lambdaFunction: lambda.Version;
  readonly eventType: cf.LambdaEdgeEventType;
  constructor(scope: Construct, id: string) {
    super(scope, id );
    const func = new lambdanj.NodejsFunction(this, 'DefaultDirIndexFunc', {
      runtime: lambda.Runtime.NODEJS_12_X,
      entry: path.join(__dirname, './lambda/cf-default-dir-index/index.ts'),
    });

    this.eventType = cf.LambdaEdgeEventType.ORIGIN_REQUEST,
    this.lambdaFunction = func.currentVersion;
  }
};