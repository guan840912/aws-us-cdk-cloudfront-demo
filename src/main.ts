import * as fs from 'fs';
import * as path from 'path';
import * as cf from '@aws-cdk/aws-cloudfront';
import * as s3 from '@aws-cdk/aws-s3';
import { BucketDeployment, Source } from '@aws-cdk/aws-s3-deployment';
import * as cdk from '@aws-cdk/core';
import { CustomWebAcl } from './custom-waf';
import * as constructs from './lambda-constructs';

const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: 'us-east-1',
};

const app = new cdk.App();
const stack = new cdk.Stack(app, 'awsus-dev', { env: devEnv });
const bucket = new s3.Bucket(stack, 'demoBucket', {
  autoDeleteObjects: true,
  removalPolicy: cdk.RemovalPolicy.DESTROY,
  websiteIndexDocument: 'index.html',
  websiteErrorDocument: 'error.html',
});

// Create pages
fs.mkdirSync(path.join(__dirname, 'a/b/c'), { recursive: true });
fs.writeFileSync(path.join(__dirname, 'index.html'), '<h1>Hello CDK!</h1>');
fs.writeFileSync(path.join(__dirname, 'a/b/c/index.html'), '<h1>Hello CDK from a/b/c/ path!!</h1>');
fs.writeFileSync(path.join(__dirname, 'error.html'), '<h1>This is an ERROR.</h1>');
fs.writeFileSync(path.join(__dirname, '404.html'), '<h1>This is a custom 404 error page.</h1>');

// Put pages to s3 bucket
new BucketDeployment(stack, 'Deployment', {
  sources: [Source.asset(path.join(__dirname, './'))],
  destinationBucket: bucket,
  retainOnDelete: false,
});

// CloudFront OriginAccessIdentity for bucket
const originAccessIdentity = new cf.OriginAccessIdentity(stack, 'OriginAccessIdentity', {
  comment: `CloudFront OriginAccessIdentity for ${bucket.bucketName}`,
});

// Create custom error page construct.
const customErrorPage = new constructs.CustomErrorPage(stack, 'CustomErrorPage');

// Create default Index construct.
const defaultIndex = new constructs.DefaultDirIndex(stack, 'DefaultIndexPage');

// Create Custom Waf V2 Rule.
const customWebAcl = new CustomWebAcl(stack, 'CustomWebAcl');

// Cloudfront distribution
const distribution = new cf.CloudFrontWebDistribution(stack, 'Distribution', {
  originConfigs: [
    {
      s3OriginSource: {
        originAccessIdentity,
        s3BucketSource: bucket,
      },
      behaviors: [{
        isDefaultBehavior: true,
        defaultTtl: cdk.Duration.seconds(10),
        lambdaFunctionAssociations: [customErrorPage, defaultIndex],
      }],
    },
  ],
  webACLId: customWebAcl.webAcl.attrArn.toString(),
});
distribution.node.addDependency(customWebAcl);

new cdk.CfnOutput(stack, 'DistributionDomainName', {
  value: distribution.distributionDomainName,
});
app.synth();