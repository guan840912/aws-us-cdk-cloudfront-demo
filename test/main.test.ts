import '@aws-cdk/assert/jest';
import { App, Stack } from '@aws-cdk/core';
import { CustomErrorPage, DefaultDirIndex } from '../src/lambda-constructs';

test('Snapshot', () => {
  const app = new App();
  const stack = new Stack(app, 'test');
  new DefaultDirIndex(stack, 'testDefaultDirIndex');
  new CustomErrorPage(stack, 'testCustomErrorPage');
  expect(app.synth().getStackArtifact(stack.artifactId).template).toMatchSnapshot();
});