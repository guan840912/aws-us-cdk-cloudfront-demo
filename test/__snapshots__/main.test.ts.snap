// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParametersb50859776d35cd33d4ae063d983b9275240a2f0473a9b8ff9c4435d91305a2b8ArtifactHashCCD60892": Object {
      "Description": "Artifact hash for asset \\"b50859776d35cd33d4ae063d983b9275240a2f0473a9b8ff9c4435d91305a2b8\\"",
      "Type": "String",
    },
    "AssetParametersb50859776d35cd33d4ae063d983b9275240a2f0473a9b8ff9c4435d91305a2b8S3Bucket8A150BED": Object {
      "Description": "S3 bucket for asset \\"b50859776d35cd33d4ae063d983b9275240a2f0473a9b8ff9c4435d91305a2b8\\"",
      "Type": "String",
    },
    "AssetParametersb50859776d35cd33d4ae063d983b9275240a2f0473a9b8ff9c4435d91305a2b8S3VersionKeyD234914C": Object {
      "Description": "S3 key for asset version \\"b50859776d35cd33d4ae063d983b9275240a2f0473a9b8ff9c4435d91305a2b8\\"",
      "Type": "String",
    },
    "AssetParametersd2c8baa20eeb156115e88c857992d932bc9c712f49f1072903b9113f4fcbe5adArtifactHashCD9013AF": Object {
      "Description": "Artifact hash for asset \\"d2c8baa20eeb156115e88c857992d932bc9c712f49f1072903b9113f4fcbe5ad\\"",
      "Type": "String",
    },
    "AssetParametersd2c8baa20eeb156115e88c857992d932bc9c712f49f1072903b9113f4fcbe5adS3Bucket7AF70871": Object {
      "Description": "S3 bucket for asset \\"d2c8baa20eeb156115e88c857992d932bc9c712f49f1072903b9113f4fcbe5ad\\"",
      "Type": "String",
    },
    "AssetParametersd2c8baa20eeb156115e88c857992d932bc9c712f49f1072903b9113f4fcbe5adS3VersionKeyD3DC4FBF": Object {
      "Description": "S3 key for asset version \\"d2c8baa20eeb156115e88c857992d932bc9c712f49f1072903b9113f4fcbe5ad\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "testCustomErrorPageCustomFunc48D44267": Object {
      "DependsOn": Array [
        "testCustomErrorPageCustomFuncServiceRoleBC4CCDB6",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersb50859776d35cd33d4ae063d983b9275240a2f0473a9b8ff9c4435d91305a2b8S3Bucket8A150BED",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersb50859776d35cd33d4ae063d983b9275240a2f0473a9b8ff9c4435d91305a2b8S3VersionKeyD234914C",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersb50859776d35cd33d4ae063d983b9275240a2f0473a9b8ff9c4435d91305a2b8S3VersionKeyD234914C",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "testCustomErrorPageCustomFuncServiceRoleBC4CCDB6",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "testCustomErrorPageCustomFuncCurrentVersion58A291F885f9f247dd42388dcedc36a82b9ce4fd": Object {
      "Properties": Object {
        "FunctionName": Object {
          "Ref": "testCustomErrorPageCustomFunc48D44267",
        },
      },
      "Type": "AWS::Lambda::Version",
    },
    "testCustomErrorPageCustomFuncServiceRoleBC4CCDB6": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "testDefaultDirIndexDefaultDirIndexFunc598E623B": Object {
      "DependsOn": Array [
        "testDefaultDirIndexDefaultDirIndexFuncServiceRole866C1485",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersd2c8baa20eeb156115e88c857992d932bc9c712f49f1072903b9113f4fcbe5adS3Bucket7AF70871",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersd2c8baa20eeb156115e88c857992d932bc9c712f49f1072903b9113f4fcbe5adS3VersionKeyD3DC4FBF",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersd2c8baa20eeb156115e88c857992d932bc9c712f49f1072903b9113f4fcbe5adS3VersionKeyD3DC4FBF",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "testDefaultDirIndexDefaultDirIndexFuncServiceRole866C1485",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "testDefaultDirIndexDefaultDirIndexFuncCurrentVersion46C0ACB1f4b8b86ee8a03bcbe7f34ee5a390dc81": Object {
      "Properties": Object {
        "FunctionName": Object {
          "Ref": "testDefaultDirIndexDefaultDirIndexFunc598E623B",
        },
      },
      "Type": "AWS::Lambda::Version",
    },
    "testDefaultDirIndexDefaultDirIndexFuncServiceRole866C1485": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
